android/java/org/chromium/chrome/browser/notifications/BraveAdsNotificationBuilder.java

```diff
@@ -0,0 +1,272 @@
+        DisplayMetrics metrics = resources.getDisplayMetrics();
+        int size = dpToPx(metrics, WORK_PROFILE_BADGE_SIZE_DP);
+        int[] colors = new int[size * size];
+
+        // Create an immutable bitmap, so that it can not be reused for painting a badge into it.
+        Bitmap bitmap = Bitmap.createBitmap(colors, size, size, Bitmap.Config.ARGB_8888);
+
+        Drawable inputDrawable = new BitmapDrawable(resources, bitmap);
+        Drawable outputDrawable = ApiCompatibilityUtils.getUserBadgedDrawableForDensity(
+                mContext, inputDrawable, null /* badgeLocation */, metrics.densityDpi);
+
+        // The input bitmap is immutable, so the output drawable will be a different instance from
+        // the input drawable if the work profile badge was applied.
+        if (inputDrawable != outputDrawable && outputDrawable instanceof BitmapDrawable) {
+            view.setImageViewBitmap(
+                    R.id.work_profile_badge, ((BitmapDrawable) outputDrawable).getBitmap());
+            view.setViewVisibility(R.id.work_profile_badge, View.VISIBLE);
+        }
+    }
+
+    /**
+     * Scales down the maximum number of displayed lines in the body text if font scaling is greater
+     * than 1.0. Never scales up the number of lines, as on some devices the notification text is
+     * rendered in dp units (which do not scale) and additional lines could lead to cropping at the
+     * bottom of the notification.
+     *
+     * @param fontScale The current system font scaling factor.
+     * @return The number of lines to be displayed.
+     */
+    @VisibleForTesting
+    static int calculateMaxBodyLines(float fontScale) {
+        if (fontScale > 1.0f) {
+            return (int) Math.round(Math.ceil((1 / fontScale) * MAX_BODY_LINES));
+        }
+        return MAX_BODY_LINES;
+    }
+
+    /**
+     * Scales down the maximum amount of flexible padding to use if font scaling is over 1.0. Never
+     * scales up the amount of padding, as on some devices the notification text is rendered in dp
+     * units (which do not scale) and additional padding could lead to cropping at the bottom of the
+     * notification. Never scales the padding below zero.
+     *
+     * @param fontScale The current system font scaling factor.
+     * @param metrics The display metrics for the current context.
+     * @return The amount of padding to be used, in pixels.
+     */
+    @VisibleForTesting
+    static int calculateScaledPadding(float fontScale, DisplayMetrics metrics) {
+        float paddingScale = 1.0f;
+        if (fontScale > 1.0f) {
+            fontScale = Math.min(fontScale, FONT_SCALE_LARGE);
+            paddingScale = (FONT_SCALE_LARGE - fontScale) / (FONT_SCALE_LARGE - 1.0f);
+        }
+        return dpToPx(metrics, paddingScale * MAX_SCALABLE_PADDING_DP);
+    }
+
+    /**
+     * Converts a px value to a dp value.
+     */
+    private static int pxToDp(float value, DisplayMetrics metrics) {
```

Comment:
I think we should have this function in a separate util class as we can use it at different places.

Reply:
Thanks for the review!
This is from current production code, how about we do this in a follow up?

Reply:
As discussed in DM, we'll merge this one as is first.

Reply:
@yrliou yeah it should be fine. I am updating onboarding for latest release. So we need to push that change as well.